nova_vm/src/ecmascript/builtins/control_abstraction_objects/iteration/iterator_prototype.rs

```diff
@@ -34,15 +50,90 @@ impl IteratorPrototype {
         Ok(this_value.unbind())
     }
 
+    /// ### [27.1.4.7 Iterator.prototype.forEach ( procedure )](https://tc39.es/ecma262/#sec-iterator.prototype.foreach)
+    fn for_each<'gc>(
+        agent: &mut Agent,
+        this_value: Value,
+        arguments: ArgumentsList,
+        mut gc: GcScope<'gc, '_>,
+    ) -> JsResult<Value<'gc>> {
+        let nogc = gc.nogc();
+        let this_value = this_value.bind(nogc);
+        let procedure = arguments.get(0).bind(nogc);
+
+        // 1. Let O be the this value.
+        // 2. If O is not an Object, throw a TypeError exception.
+        let Ok(o) = Object::try_from(this_value) else {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "'this' is not an object",
+                nogc,
+            ));
+        };
+
+        // 3. Let iterated be the Iterator Record { [[Iterator]]: O, [[NextMethod]]: undefined, [[Done]]: false }.
+        // 4. If IsCallable(procedure) is false, then
+        let Some(procedure) = is_callable(procedure, nogc).unbind().bind(gc.nogc()) else {
+            // a. Let error be ThrowCompletion(a newly created TypeError object).
+            let error = agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "'procedure' is not callable",
+                nogc,
+            );
+            // b. Return ? IteratorClose(iterated, error).
+            return iterator_close(agent, o.unbind(), Err(error), gc);
+        };
+        let scoped_procedure = Value::from(procedure).scope(agent, gc.nogc());
+
+        // 5. Set iterated to ? GetIteratorDirect(O).
+        // if i rebind `iterated` here borrow checker complains
+        let iterated = get_iterator_direct(agent, o.unbind(), gc.reborrow()).unbind()?;
```

Comment:
Binding `iterated` here results in both the reborrow calls in the loop saying "cannot borrow `gc` as mutable because it is also borrowed as immutable`. I can't scope the iterator either so I just left it as unbound which seems to fix it. This is probably wrong, though.

Reply:
Yeah, it'll eventually lead to UAF once we do GC.

Now that the `done` field is gone from the iterator record, it's equal to just two bound values. So what I've done elsewhere is to take out the two bound values and scoped them separately, then I recreate the IteratorRecord at every callsite. It's not pretty, but it works and is probably even fairly optimal.